File: test/java/src/test/java/org/tinyspline/InterpolationTest.java
package org.tinyspline;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.function.Predicate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.tinyspline.tinysplinejavaConstants.TS_POINT_EPSILON;

public class InterpolationTest {

    private String str = "";

    private Predicate<? super Vec2> dist(double x, double y) {
        Vec2 vec2 = new Vec2(x, y);
        str = "distance => " + vec2;
        return vec -> vec.distance(vec2) <= TS_POINT_EPSILON;
    }

    @Test
    public void testCubicNatural() {
        // Given
        List<Double> points = List.of(
                1.0, -1.0, // P1
                -1.0, 2.0, // P2
                1.0, 4.0,  // P3
                4.0, 3.0,  // P4
                7.0, 5.0   // P5
        );

        // When
        BSpline spline = BSpline.interpolateCubicNatural(points, 2);

        // Then
        assertThat(spline.getNumControlPoints()).isEqualTo(16);

        // First bezier segment
        assertThat(spline.controlPointVec2At(0)).matches(dist(1, -1), str);
        assertThat(spline.controlPointVec2At(1)).matches(dist(0, 0), str);
        assertThat(spline.controlPointVec2At(2)).matches(dist(-1, 1), str);
        assertThat(spline.controlPointVec2At(3)).matches(dist(-1, 2), str);

        // Second bezier segment
        assertThat(spline.controlPointVec2At(4)).matches(dist(-1, 2), str);
        assertThat(spline.controlPointVec2At(5)).matches(dist(-1, 3), str);
        assertThat(spline.controlPointVec2At(6)).matches(dist(0, 4), str);
        assertThat(spline.controlPointVec2At(7)).matches(dist(1, 4), str);

        // Third bezier segment
        assertThat(spline.controlPointVec2At(8)).matches(dist(1, 4), str);
        assertThat(spline.controlPointVec2At(9)).matches(dist(2, 4), str);
        assertThat(spline.controlPointVec2At(10)).matches(dist(3, 3), str);
        assertThat(spline.controlPointVec2At(11)).matches(dist(4, 3), str);

        // Forth bezier segment
        assertThat(spline.controlPointVec2At(12)).matches(dist(4, 3), str);
        assertThat(spline.controlPointVec2At(13)).matches(dist(5, 3), str);
        assertThat(spline.controlPointVec2At(14)).matches(dist(6, 4), str);
        assertThat(spline.controlPointVec2At(15)).matches(dist(7, 5), str);
    }
}

File: test/csharp/InterpolationTest.cs
using NUnit.Framework;
using System.Collections.Generic;
using TinySpline;

namespace csharp;

public class InterpolationTest
{
    class Vec2EqualityComparer : IEqualityComparer<Vec2>
    {
        public bool Equals(Vec2? v1, Vec2? v2)
        {
            if (v1 == null && v2 == null) return true;
            if (v1 == null || v2 == null) return false;
            return v1.Distance(v2) <= tinysplinecsharp.TS_POINT_EPSILON;
        }

        public int GetHashCode(Vec2 vec)
        {
            double hCode = vec.X * vec.Y;
            return hCode.GetHashCode();
        }
    }

    [Test]
    public void TestCubicNatural()
    {
        // Given
        List<double> points = new List<double> {
            1.0, -1.0, // P1
            -1.0, 2.0, // P2
            1.0, 4.0,  // P3
            4.0, 3.0,  // P4
            7.0, 5.0   // P5
        };
        Vec2EqualityComparer comparer = new Vec2EqualityComparer();

        // When
        BSpline spline = BSpline.InterpolateCubicNatural(points, 2);

        // Then
        Assert.That(spline.NumControlPoints, Is.EqualTo(16));

        // First bezier segment
        Assert.That(spline.ControlPointVec2At(0),
                    Is.EqualTo(new Vec2(1, -1))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(1),
                    Is.EqualTo(new Vec2(0, 0))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(2),
                    Is.EqualTo(new Vec2(-1, 1))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(3),
                    Is.EqualTo(new Vec2(-1, 2))
                      .Using(comparer));

        // Second bezier segment
        Assert.That(spline.ControlPointVec2At(4),
                    Is.EqualTo(new Vec2(-1, 2))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(5),
                    Is.EqualTo(new Vec2(-1, 3))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(6),
                    Is.EqualTo(new Vec2(0, 4))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(7),
                    Is.EqualTo(new Vec2(1, 4))
                      .Using(comparer));

        // Third bezier segment
        Assert.That(spline.ControlPointVec2At(8),
                    Is.EqualTo(new Vec2(1, 4))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(9),
                    Is.EqualTo(new Vec2(2, 4))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(10),
                    Is.EqualTo(new Vec2(3, 3))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(11),
                    Is.EqualTo(new Vec2(4, 3))
                      .Using(comparer));

        // Forth bezier segment
        Assert.That(spline.ControlPointVec2At(12),
                    Is.EqualTo(new Vec2(4, 3))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(13),
                    Is.EqualTo(new Vec2(5, 3))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(14),
                    Is.EqualTo(new Vec2(6, 4))
                      .Using(comparer));
        Assert.That(spline.ControlPointVec2At(15),
                    Is.EqualTo(new Vec2(7, 5))
                      .Using(comparer));
    }
}
